### ISSUE TITLE
NSS integration tasks

### ISSUE THREAD

**Issue by beurdouche** [mitls/hacl-star#20]
General improvements (required)
[ ] Generate new snapshot with parenthesis to silence `-Werror`

Additional algorithms and changes (if possible)
[ ] Curve25519 (32bits)
[ ] SHA2 incremental with standard interface

**Comment by beurdouche** [comment]
Curve25519 (64bits)
https://bugzilla.mozilla.org/show_bug.cgi?id=1376975
C1 - https://github.com/beurdouche/nss/commit/4c176f62f16fc0d13c2f9229bd8d235b43e4ab02
C2 - https://github.com/beurdouche/nss/commit/c2934a933cc7a6d2c4e807f845bb40f3763b553f
Missing Licence. Waiting on legal to see if Apache2 is possible

**Comment by beurdouche** [comment]
```
SHA2, HMAC, HKDF
```
- [ ] Rewrite with the standard interface
- [ ] Cleanup the extracted code
- [ ] Cleanup the following pattern by not removing a tuple 
```
  if (len < (uint32_t )56)
    uu____1971 = ((K___uint32_t_uint8_t_ ){ .fst = (uint32_t )1, .snd = blocks + (uint32_t )64 });
  else
    uu____1971 = ((K___uint32_t_uint8_t_ ){ .fst = (uint32_t )2, .snd = blocks });
  K___uint32_t_uint8_t_ scrut = uu____1971;
```

**Comment by franziskuskiefer** [comment]
- [ ] The generated code has inlined functions. This will make problems when building on Windows. I think the compiler is intelligent enough to inline. Can we drop the inlines?
- [ ] Some void functions have returns (not all). Can we not do that?

**Comment by msprotz** [comment]
@franziskuskiefer can you elaborate on the first point? I know that JK, Benjamin and others specifically requested that the kremlin tool inline some functions at call-site for performance, rather than slapping a C `inline` annotation on the function.
- Are you asking for the latter behavior?
- What sort of problems do you foresee on windows?

Thanks!

**Comment by karthikbhargavan** [comment]
I am also curious about this.
The reason for some of the “inline” annotations is that they do seem to be needed by GCC to produce efficient code, although some of them may be unnecessary.
What compiler would you use on windows?
@protz, have we tried compiling hacl* with MSVC?

-K.

**Comment by msprotz** [comment]
Yes, we run CI on Windows and until recently we checked that the extracted files did compile with MSVC. It looks like now, we only check that kremlin extracts everything properly when passed `-cc msvc`.

@beurdouche any way we can restore compilation of the test files as part of the extract-c target?

**Comment by beurdouche** [comment]
@protz Yes, I think we should do that. Compiling under GCC, MSVC and CompCert as part of the CI would be nice, and it is not very time consuming compared to verification... (Put that in the bucket list)

**Comment by franziskuskiefer** [comment]
`inline` doesn't exist on [Windows in C](https://msdn.microsoft.com/en-us/library/bw1hbe6y.aspx).

> The inline keyword is available only in C++. The __inline and __forceinline keywords are available in both C and C++. For compatibility with previous versions, _inline is a synonym for __inline.

So I think NSS avoids using `inline` (after all it's only a hint). I'm sure we could get `inline` working, but it would need some `#ifdef` to make it work. If we don't want to do that and the code gets less efficient when we take out `inline`, I'd suggest inlining the functions where possible.

**Comment by msprotz** [comment]
The commit above removes the occurrences of `(void) foo`.

**Comment by msprotz** [comment]
The commit above fixes the superfluous `return`s in void-returning functions.

**Comment by msprotz** [comment]
Adding @tahina-pro for " MSVC, GCC, CompCert compilation of the snapshot" under the CI section

**Comment by franziskuskiefer** [comment]
We have to add a license header for the generated code (also see #17). Apache 2 would work for NSS.
